Fix various issues on the Dashboard

somebody1234 opened the pull request:
### Pull Request Description
- Fix React DevTools not working in Firefox
- Fix selection of asset names when editing them, not working at all in Firefox
- Convert tick/cross buttons when editing assets, and the "plus" and "reload" buttons on the "shared with" column, "labels" column, and keyboard shortcuts table, to match more with the rest of the design
- Update clip path when the container resizes, so that the icons for hidden columns never overlap the actual table header
- Fix #10184
  - Fix renames being committed even when cancelling
- Fix duplicate name detection - previously, all asset types only checked folders with the same name, not assets with the same name
  - I'm not 100% sure this is the correct behavior still
- Stop using `kbd` (`aria.Keyboard`) to display keyboard shortcuts, since they should not be displayed in a monospace font.
- Fix "plus" and "reload" buttons going past the right side of their parent table cell
- Limit length of `PermissionDisplay` - if the username of a user with permission is too long, it uses a tooltip instead
- Update the username dynamically for all permissions owned by self, when changing username in the settings.
  - This avoids having to fully invalidate the directory tree every time the username changes, given that nothing changes about the assets' metadata themselves.
- Cache children in the Drive tree
  - This avoids loading spinners when closing a folder and immediately reopening it.
  - Note that children are still re-fetched on reopen to ensure freshness

### Important Notes
- This MAY be split into multiple smaller PRs. However, I think it's better to QA as a single PR, to avoid duplicating work checking behavior that may be changed by a sibling PR (assuming the PR was split into multiple).

### Testing Instructions
- Mostly test that the above bugs have been fixed
- But also test the rest of the app in general, because quite a few things *are* being changed here

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] ~~The documentation has been updated, if necessary.~~
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] ~~Unit tests have been written where possible.~~

somebody1234 committed:
Finally fix React DevTools being broken...

somebody1234 committed:
Clean up code

somebody1234 committed:
Change `object.merge` and `object.merger` to reduce unnecessary re-renders in order to aid with debugging

somebody1234 committed:
Cache directory contents in Asset Table

somebody1234 committed:
Further debugging and fixes for editing name...

somebody1234 committed:
WIP: still debugging broken text selection...

somebody1234 committed:
Avoid caching directory entries

somebody1234 committed:
Remove debug logs

somebody1234 committed:
Fix Firefox selection bug; document browser-specific behavior

somebody1234 committed:
Minor rename of `DragModal` props

somebody1234 committed:
Replace disabled `exhaustive-hooks` lints with refs

somebody1234 committed:
Option to limit `SelectionBrush` to a given element

somebody1234 committed:
Remove `/* should never change */` annotations (they were unnecessary)

somebody1234 committed:
Update username live for asset permissions when changing username

somebody1234 committed:
Limit `PermissionDisplay` length and show tooltip when too long

somebody1234 committed:
Fix flex containers being too small for buttons in table cells

somebody1234 committed:
Update design of buttons to be more consistent with other icon buttons

somebody1234 committed:
Fix E2E tests

somebody1234 committed:
Move classes to `tailwind.css`

somebody1234 committed:
Use `apply` instead of hardcoded CSS rules

somebody1234 committed:
WIP: Improve button styles

somebody1234 committed:
Fix folders collapsing when refocusing window

somebody1234 committed:
Stop using `kbd` to display keyboard shortcuts

somebody1234 committed:
Add `useOnScroll` to `exhaustive-deps` lint

somebody1234 committed:
Update `AssetsTable` clip path on resize using `ResizeObserver`

somebody1234 committed:
More consistent buttons

somebody1234 committed:
Switch buttons to `ariaComponents.Button`

somebody1234 committed:
Adjust styles for buttons; adjust styles for plus button in "shared with" and "labels" columns

somebody1234 committed:
Convert tick and cross icons to be consistent with other icons

somebody1234 committed:
Fix cancelling rename not working

somebody1234 committed:
Fix duplicate name detection

somebody1234 committed:
Add more validation for directory names

MrFlashAccount commented on app/ide-desktop/lib/dashboard/src/utilities/Debug.tsx:
```diff
@@ -0,0 +1,60 @@
+    console.group(`[Debug(${typeName})] Rendering`, Component, patchedChildProps)
+  }
+  const element = <Component {...patchedChildProps} />
+  if (monitorRender) {
```
Render doesn't finish after calling `<Component {...patchedChildProps} />`

somebody1234 replied:
oops, true... i'm not sure whether it's useful to track renders, so i'm thinking of just renaming it from `Rendering` to `Constructing React Node` or similar. thoughts?

somebody1234 replied:
i guess i'll test with nesting `Debug` and do whatever makes the most sense with that

somebody1234 replied:
should be fixed now?

MrFlashAccount replied:
> Constructing React Node or similar. thoughts

For me, it makes no sense... constructing a node is a trivial operation and gives no debug info at all.

MrFlashAccount commented on app/ide-desktop/lib/dashboard/src/utilities/Debug.tsx:
```diff
@@ -0,0 +1,60 @@
+/** @file Utilities related to debugging. */
+import * as React from 'react'
```
I have no idea why we have this file. What feature of current chrome/react devtools we are missing so do we need to add a component?

somebody1234 replied:
this provides a convenient way to track changed props for a single element, IME it's quite difficult to filter renders to just the ones that affect a specific component, and even more difficult to find out the props that changed, let alone the values of the changed props

(oh, also it logs when functions are called - i was trying to figure out whether some drag handler was causing issues on Firefox)

MrFlashAccount replied:
There's a flame graph feature: 
![CleanShot 2024-06-16 at 16 46 29](https://github.com/enso-org/enso/assets/61194245/89a95803-9b48-4d35-bd1f-0b893b2bc72a)

Maybe this can help?

somebody1234 replied:
i do indeed use react devtools, but the flame graph is, in fact, not useful for determining which functions were called - given that the flame graph shows the component tree, not the call stack. and i find the render reasons to be cryptic, and missing the call stack of the original function that triggered the rerender in the first place.

not to mention, the issue here is that the bug was occurring even when *no* renders were happening, so react devtools could not possibly be useful in the first place.

MrFlashAccount replied:
Oh, that makes sense. thanks for the clarification.

I still doubt we need this component because we might have forgotten to remove usage from the codebase and debug logs will appear in production ... yikes!

somebody1234 replied:
should be easy enough to add a lint rule to make it fail CI checks... thoughts?

MrFlashAccount replied:
I'm fine with that

somebody1234 committed:
WIP: playing with `Debug`

somebody1234 committed:
Fix `tertiary` buttons

somebody1234 committed:
Fix output of `Debug` component

somebody1234 committed:
Merge branch 'develop' into wip/sb/dashboard-fixes

somebody1234 committed:
Fix lint errors

somebody1234 committed:
Fix duplicate name detection

somebody1234 committed:
Fix E2E tests

somebody1234 committed:
Minor CSS changes

somebody1234 committed:
Prettier

somebody1234 committed:
Remove `shouldPreventNavigation`

MrFlashAccount commented on app/ide-desktop/lib/dashboard/src/hooks/debugHooks.ts:
```diff
@@ -49,25 +49,28 @@ export function useDebugState<T>(
 
 /** A helper function to log the old and new values of changed dependencies. */
-function useMonitorDependencies(
+export function useMonitorDependencies(
```
Are we just tracking changes between renders, right?

somebody1234 replied:
yes, for most of the debug hooks it logs the reason why the hook is being re-run.
it *is* used in `Debug.tsx` but is totally optional there and can be removed.

(or if you mean the documentation isn't quite descriptive enough then that can be improved too.)

MrFlashAccount replied:
Btw did we include this hook into eslint aswell?

somebody1234 replied:
which eslint rule? there is one for callback dependencies but that doesn't apply here as this hook doesn't accept a callback

MrFlashAccount replied:
The same as for the `<Debug />` component: https://github.com/enso-org/enso/pull/10256#discussion_r1641885583

MrFlashAccount commented on app/ide-desktop/lib/dashboard/src/components/AriaComponents/Button/Button.tsx:
```diff
@@ -78,6 +78,7 @@ export const BUTTON_STYLES = twv.tv({
     'transition-[opacity,outline-offset,background,border-color] duration-150 ease-in-out',
   ],
   variants: {
+    isCentered: { true: '' },
```
What's the purpose of this prop? I'm starting being worry about the amount of boolean props for a button.

somebody1234 replied:
without `isCentered`, the icon buttons with an outline are slightly off center because of the padding used to center the text. definitely far from an ideal solution, but i just did whatever was the least likely to break all other buttons - if you have a better solution i'd be happy to switch

MrFlashAccount replied:
Am I right that this is happening when the button contains only an icon?

somebody1234 replied:
yup - see here: <https://github.com/enso-org/enso/pull/10256/files/8f11e9e5582d4cd04dba940eaca6addc05ace307#diff-3fe65ff23a2e634db1d9cb4ea510d11e7450db13a58c2b752c7c3cfc3fa8af47R239-R243>

somebody1234 replied:
oh, i guess one other thing to note is that the compound variant is also used as a hack to get an icon with an exact (16px in this case) pixel size, to avoid icons being more blurry than they need to be. not sure what's a good way to solve that issue

MrFlashAccount replied:
Oh, I see, then the problem is with `isIconOnly` prop. Currently, we customize paddings when icon is preset only in `icon` variant. Maybe it's time to extend this behavior?

somebody1234 replied:
well, i'm not sure the problem is with the `isIconOnly` prop - rather the issue is the `h-[2.5cap] -mt-[0.1cap]` classes that are applied, causing the icons to be (apparently) off center when an outline is applied, and also causing the icons to be not 16px.

as mentioned above, there's probably a better way to solve this, but i picked the option that was least likely to break some button somewhere i might not know about

MrFlashAccount replied:
> least likely to break some button somewhere i might not know about

Don't worry, not a huge deal. it's better to break a button visually(a bit though), than introduce a new option(which increases complexity)

somebody1234 replied:
(to elaborate: iirc it *was* using `isIconOnly`, and i think that was the exact thing that was causing this issue.)

see example usages here for example:
https://github.com/enso-org/enso/pull/10256/files/8f11e9e5582d4cd04dba940eaca6addc05ace307#diff-f599366d794555bea0bd1d9563b890e56eb4a692687fd1ce8e4fcd9a1d22902bR114-R117

without the explicit `isCentered` prop, i think it would fallback to `isIconOnly`

(i do see that there *is* a dedicated `icon` variant, but it doesn't seem to affect the `icon` slot?)

somebody1234 committed:
Add ESLint rule to ban `<Debug>` component

PabloBuchu commented:
| Limit length of PermissionDisplay - if the username of a user with permission is too long, it uses a tooltip instead
1. How long is too long?
<img width="718" alt="Screenshot 2024-06-18 at 12 23 04" src="https://github.com/enso-org/enso/assets/2855109/4139b7cb-621b-4457-aedb-0d904a2dcb61">

2. When dropping a label on directory PATCH request is being send twice (at least on Firefox)
3. Reload on firefox works selectively (sometimes I need to reload 2 or 3 times or hard reload for dashboard to apear)
4. Removing labels is not working on Firefox
5. When I want to drop a file on dashboard I see white box

somebody1234 committed:
Use `IntersectionObserver` to determine whether to show secondary dropzone

somebody1234 committed:
Refactor and fix hook to determine whether to show backup dropzone

somebody1234 committed:
Fix truncation of permission pills

somebody1234 commented:
1. and 5. addressed, will address the others soon

somebody1234 committed:
Fix weird behavior when dragging labels...

somebody1234 committed:
Improve lag when dragging labels on Firefox

somebody1234 commented:
ok i *think* 2 and 4 are now addressed as well. for 3 i think i might need more details - what happens when reloading fails, instead of the dashboard appearing? (screenshots and screencasts would be useful)

somebody1234 commented:
(@PabloBuchu i guess this PR is ready for re-QA?)

somebody1234 committed:
Merge branch 'develop' into wip/sb/dashboard-fixes

somebody1234 committed:
Fix permission selector appearance

somebody1234 committed:
Remove `isCentered` prop in favor of custom size

somebody1234 committed:
Fix E2E tests

somebody1234 committed:
Use custom `tailwind-merge` config for `tailwind-variants`

somebody1234 committed:
Use non-custom size for Project icons

somebody1234 committed:
Fix infinite loop in `AssetProperties`

somebody1234 committed:
Fix `package.json`

somebody1234 committed:
Merge branch 'develop' into wip/sb/dashboard-fixes

somebody1234 committed:
Unselect current page when switching to settings page

somebody1234 committed:
Truncate username and org name on settings page

MrFlashAccount committed:
Fix button in kb shortcuts

somebody1234 committed:
Fix button styling

somebody1234 committed:
Hide Keyboard Shortcuts settings page buttons unless `focus-within` or hovered over

somebody1234 committed:
Fix buttons on settings pages

MrFlashAccount commented:
QA 1 ❌

1. UserName/OrgName in the title is wider than it should be. Also, I'm not sure it's a good idea to display a tooltip if the content doesn't overflow(maybe it's due to hardcoded height, told ya to not specify height explicitly) 
![CleanShot 2024-06-20 at 11 55 18](https://github.com/enso-org/enso/assets/61194245/3e6c5e68-55d1-4b07-bcfd-2ae97c9f762d)
2. On my account, the Name input in settings is empty: 
![CleanShot 2024-06-20 at 11 59 22](https://github.com/enso-org/enso/assets/61194245/e8f4e9c2-1057-48c4-96d3-6ee8b268ddb9)
3. KBD shortcuts are not aligned 
![CleanShot 2024-06-20 at 12 00 16](https://github.com/enso-org/enso/assets/61194245/0e1c9258-9dd9-4dba-a314-428cba252c69)
4. (Not sure it's related to this PR, can be a separate issue) When I have a long project name and want to rename it, the table resizes and I might misclick on the input and renaming discards(see video attached)

https://github.com/enso-org/enso/assets/61194245/a84ac55c-ef5d-4c58-80f6-9ba8477853b0

5. (Not sure it's related to this PR, can be a separate issue) We should trigger rename by clicking anywhere on the name field, not only on the text
6.  (Not sure it's related to this PR, can be a separate issue) When I quickly double-click on the name field, renaming doesn't show up.
7. (This can be a separate issue) Can we also limit the length of labels as well? 
![CleanShot 2024-06-20 at 12 09 03](https://github.com/enso-org/enso/assets/61194245/6cf4f5e6-deb7-4d26-9831-5acf2cddd27b)

MrFlashAccount commented on app/ide-desktop/lib/dashboard/src/components/AriaComponents/Button/Button.tsx:
```diff
@@ -96,6 +96,8 @@ export const BUTTON_STYLES = twv.tv({
       custom: { base: '', extraClickZone: '', icon: 'h-full' },
       icon: { icon: 'h-4' },
+      // eslint-disable-next-line @typescript-eslint/naming-convention
+      'project-icon': { icon: 'h-6' },
```
I disagree with this approach. The button should be unaware of where it's used, thus sizing shouldn't contain the _exact place_-related tokens. If we want to introduce another size we should fit this into the current sizing model, when it's clear how sizes relate to each other: 'small', 'medium', 'large', etc. This naming has infinite growing potential, but I'd suggest you avoid adding new sizes unless it's strictly necessary.

MrFlashAccount commented on app/ide-desktop/lib/dashboard/src/components/EditableSpan.tsx:
```diff
@@ -122,32 +124,30 @@ export default function EditableSpan(props: EditableSpanProps) {
+        <ariaComponents.ButtonGroup gap="xsmall" className="grow-0 items-center">
+          {isSubmittable && (
+            <ariaComponents.Button
+              size="icon"
```
same here, size=icon doesn't show the relative size of the element.
But I understand why you did that. Let's leave it for now as a technical debt

MrFlashAccount commented on app/ide-desktop/lib/dashboard/src/components/JSONSchemaInput.tsx:
```diff
@@ -49,14 +51,10 @@ export default function JSONSchemaInput(props: JSONSchemaInputProps) {
-    setValueRaw(value)
-    // `setStateRaw` is a callback, not a dependency.
-    // eslint-disable-next-line react-hooks/exhaustive-deps
+    setValueRawRef.current(value)
```
I assume "useEffect" might be too late here. Since we're assigning to a ref, we can safely do this during the render. Please refer to https://react.dev/learn/you-might-not-need-an-effect to learn more about patterns you can use to avoid using "useEffect" 

P.S. We have `useSyncRef` hook(app/ide-desktop/lib/dashboard/src/hooks/syncRefHooks.ts) that does exactly what you want

MrFlashAccount commented on app/ide-desktop/lib/dashboard/src/components/SelectionBrush.tsx:
```diff
@@ -11,129 +11,148 @@ import * as eventModule from '#/utilities/event'
+      }
+    }
+
+    document.addEventListener('mousedown', onMouseDown)
```
Can we return `target` here instead of `document`? this was intended - we want to start selection only on the particular element

MrFlashAccount commented on app/ide-desktop/lib/dashboard/src/components/dashboard/DirectoryNameColumn.tsx:
```diff
@@ -188,16 +190,15 @@ export default function DirectoryNameColumn(props: DirectoryNameColumnProps) {
-              child.item.title !== newTitle
-          )
+          validation.DIRECTORY_NAME_REGEX.test(newTitle) &&
+          (nodeMap.current.get(item.directoryKey)?.children ?? []).every(child => {
```
looks like a copypaste, can we avoid that here?

MrFlashAccount commented on app/ide-desktop/lib/dashboard/src/components/dashboard/column/LabelsColumn.tsx:
```diff
@@ -141,10 +141,11 @@ export default function LabelsColumn(props: column.AssetColumnProps) {
-          className="shrink-0 rounded-full opacity-0 group-hover:opacity-100 focus-visible:opacity-100"
+          size="icon"
+          variant="outline"
+          className="opacity-0 group-hover:opacity-100 focus-visible:opacity-100"
```
AFAIR `group` here relates to this button, doesn't it?

somebody1234 replied:
if you mean the `group` in the ancestor `div`, then yes

MrFlashAccount replied:
Button creates its own group: https://github.com/enso-org/enso/blob/ccfeac1c02b3f6637f1e66650d712afb1f175af1/app/ide-desktop/lib/dashboard/src/components/AriaComponents/Button/Button.tsx#L69

I'm not sure how tw will resolve `group-*` selector in this situation.

somebody1234 replied:
ah, good to know. i think it's not a problem as `.group` matches any group:

![image](https://github.com/enso-org/enso/assets/4046547/f693a23a-68f2-411e-aacc-baf082e523cf)

and all we care about is that the button is visible when the `group` is hovered/contains focus.

somebody1234 commented:
> We should trigger rename by clicking anywhere on the name field, not only on the text

this should already be the case i think

> When I quickly double-click on the name field, renaming doesn't show up

double clicking opens the asset (projects are run, directories are toggled).

the mouse shortcut for rename is supposed to be single click + single click - apparently this is the behavior of finder on macos?


will definitely address the other issues though

somebody1234 committed:
Address QA

MrFlashAccount commented on app/ide-desktop/lib/dashboard/src/hooks/useOnScroll.ts:
```diff
@@ -0,0 +1,43 @@
+
+/** Execute a function on scroll. */
+export default function useOnScroll(callback: () => void, dependencies: React.DependencyList) {
+  const callbackRef = React.useRef(callback)
```
btw, we have `useEventCallback` app/ide-desktop/lib/dashboard/src/hooks/eventCallbackHooks.ts that does exactly this.

MrFlashAccount commented on app/ide-desktop/lib/dashboard/src/layouts/Settings.tsx:
```diff
@@ -110,16 +114,22 @@ export default function Settings(props: SettingsProps) {
+        <ariaComponents.Text
+          variant="h1"
+          truncate="1"
+          className="ml-[0.625rem] h-[2.25rem] max-w-[30ch] rounded-full bg-frame px-[0.5625rem] pb-[0.3125rem] pt-[0.125rem] leading-snug"
```
Btw, would like to take a look at this change? https://github.com/enso-org/enso/pull/10171/files#diff-b4033a983460df383c4932fac422821a9066585230aedee437e2540f2e4da514R146 

I did it for the Billing page(it also fixes nasty overflowing bugs in settings). Maybe we can cherry pick these changes instead?

MrFlashAccount commented:
> double clicking opens the asset (projects are run, directories are toggled).

Hm, just tried - nothing happend

somebody1234 committed:
Address CR

somebody1234 committed:
Refactor name check

somebody1234 committed:
Merge branch 'develop' into wip/sb/dashboard-fixes

somebody1234 committed:
Fix double clicking project to open

somebody1234 commented:
1. addressed
2. addressed
3. can't repro (after temporarily changing linux modifier keys to icons) - it may be easier for you to fix it?
4. addressed
5. N/A (see above)
6. N/A (see above)
7. addressed

CRs should be addressed as well

somebody1234 committed:
Auto-resize `EditableSpan` for asset name input

somebody1234 commented:
4. also addressed

somebody1234 committed:
Fix E2E tests

MrFlashAccount committed:
Fix keyboard shortcuts

MrFlashAccount commented:
QA ✅

somebody1234 committed:
Merge branch 'develop' into wip/sb/dashboard-fixes
